src/vision/status_quo/barbara_bridges_sync_and_async.md

```diff
@@ -0,0 +1,245 @@
+        .iter()
+        .map(|url| block_on(do_web_request(url)))
+        .collect()
+}
+```
+
+The good news is that the deadlock is gone. The bad news is that now she is getting a panic:
+
+> thread 'main' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.'
+
+"Well," she thinks, "I could use that `Handle` API to get the current runtime, maybe that will work?"
+
+```rust
+fn aggregate(urls: &[&str]) -> Vec<String> {
+    let handle = Handle::current();
+    urls.iter()
+        .map(|url| handle.block_on(do_web_request(url)))
+        .collect()
+}
+```
+
+But it seems to give her the same panic. 
+
+### Trying out `spawn_blocking`
+
+Reading more into this problem, she realizes she is supposed to be using `spawn_blocking`. She tries replacing `block_on` with `tokio::task::spawn_blocking`:
+
+```rust=
+fn aggregate(urls: &[Url]) {
+    let data: Vec<Data> = 
+        urls
+            .iter()
+            .map(|url| tokio::task::spawn_blocking(move || do_web_request(url)))
+            .collect();
+}
+```
+
+
+but now she gets a type error again:
+
+```
+error[E0277]: a value of type `Vec<Data>` cannot be built from an iterator over elements of type `tokio::task::JoinHandle<impl futures::Future>`
+  --> src/main.rs:22:14
+   |
+22 |             .collect();
+   |              ^^^^^^^ value of type `Vec<Data>` cannot be built from `std::iter::Iterator<Item=tokio::task::JoinHandle<impl futures::Future>>`
+   |
+   = help: the trait `FromIterator<tokio::task::JoinHandle<impl futures::Future>>` is not implemented for `Vec<Data>`
+```
+
+Of course! `spawn_blocking`, like `map`, just takes a regular closure, not an async closure. She's getting a bit frustrated now. "Well," she thinks, "I can use `spawn` to get into an async context!" So she adds a call to `spawn` inside the `spawn_blocking` closure:
+
+```rust
+fn aggregate(urls: &[Url]) {
+    let data: Vec<Data> =
+        urls
+            .iter()
+            .map(|url| tokio::task::spawn_blocking(move || {
+                tokio::task::spawn(async move {
+                    do_web_request(url).await
+                })
+            }))
+            .collect();
+}
+```
+
+But this isn't really helping, as `spawn` still yields a future. She's getting the same errors.
+
+### Async all the way
+
+She remembers now that this whole drama started because she was converting her `main` function to be `async`. Maybe she doesn't have to bridge between sync and async? She starts digging around in the docs and finds `futures::join_all`. Using that, she can change `aggregate` to be an async function too:
+
+```rust
+async fn aggregate(urls: &[Url]) {
+    let data: Vec<Data> = futures::join_all(
+        urls
+            .iter()
+            .map(|url| do_web_request(url))
+    ).await;
+}
+```
+
+Things are working again now, so she is happy.
```

Comment:
I wanted to leave one more final solution to which she might reach.

```rust
fn aggregate(urls: &[Url], handle: Handle) -> Vec<Data> {
    urls.iter()
        .map(|url| handle.block_on(do_web_request(url)))
        .collect()
}

#[tokio::main]
async fn main() {
    let handle = Handle::current();
    let data = task::spawn_blocking(move || aggregate(&[Url, Url], handle))
        .await
        .unwrap();
}
```

Reply:
In some cases, `aggregate` function may not be changed (happened to me). She may not have the luxury to pass the handle to the child function all the time if a third party crate was involved.

Reply:
She could use `Handle::current` from inside `spawn_blocking`, I believe. Question is, would that actually work? What would happen if she did that? I guess I can try it myself...

Reply:
[It seems to work fine](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bac20beedbe6eacc0a2a6d7a4d993e8b)

Reply:
Yes, we guarantee that `Handle::current()` works fine in `spawn_blocking`.

Reply:
I wanted to leave one more thing here, `Handle::block_on(...)` was introduced in tokio 1.4. And this solution is definitely possible, but to reach till here, might require folks to read the documentation thoroughly.

Reply:
Added a FAQ with this variant.